mongodbatlas/fw_resource_mongodbatlas_database_user.go

```diff
@@ -0,0 +1,529 @@
+						"value": schema.StringAttribute{
+							Optional: true,
+							Computed: true,
+						},
+					},
+				},
+			},
+			"scopes": schema.SetNestedBlock{
+				NestedObject: schema.NestedBlockObject{
+					Attributes: map[string]schema.Attribute{
+						"name": schema.StringAttribute{
+							Optional: true,
+						},
+						"type": schema.StringAttribute{
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func NewDatabaseUserRS() resource.Resource {
+	return &DatabaseUserRS{}
+}
+
+func (r *DatabaseUserRS) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = fmt.Sprintf("%s_%s", req.ProviderTypeName, databaseUserResourceName)
+}
+
+func (r *DatabaseUserRS) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	client, err := ConfigureClientInResource(req.ProviderData)
+	if err != nil {
+		resp.Diagnostics.AddError(errorConfigureSummary, err.Error())
+		return
+	}
+	r.client = client
+}
+
+func (r *DatabaseUserRS) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var databaseUserPlan *tfDatabaseUserModel
+
+	diags := req.Plan.Get(ctx, &databaseUserPlan)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	dbUserReq, d := newMongoDBDatabaseUser(ctx, databaseUserPlan)
+	resp.Diagnostics.Append(d...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	conn := r.client.Atlas
+	dbUser, _, err := conn.DatabaseUsers.Create(ctx, databaseUserPlan.ProjectID.ValueString(), dbUserReq)
+	if err != nil {
+		resp.Diagnostics.AddError("error during database user creation", err.Error())
+		return
+	}
+
+	dbUserModel, diagnostic := newTFDatabaseUserModel(ctx, databaseUserPlan, dbUser)
+	resp.Diagnostics.Append(diagnostic...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &dbUserModel)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+func (r *DatabaseUserRS) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var databaseUserState *tfDatabaseUserModel
+	var err error
+	resp.Diagnostics.Append(req.State.Get(ctx, &databaseUserState)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	username := databaseUserState.Username.ValueString()
+	projectID := databaseUserState.ProjectID.ValueString()
+	authDatabaseName := databaseUserState.AuthDatabaseName.ValueString()
+
+	// Use the ID only with the IMPORT operation
+	if databaseUserState.ID.ValueString() != "" && (username == "" || projectID == "" || authDatabaseName == "") {
+		projectID, username, authDatabaseName, err = splitDatabaseUserImportID(databaseUserState.ID.ValueString())
+		if err != nil {
+			resp.Diagnostics.AddError("error splitting database User info from ID", err.Error())
+			return
+		}
+	}
+
+	conn := r.client.Atlas
+	dbUser, _, err := conn.DatabaseUsers.Get(ctx, authDatabaseName, projectID, username)
```

Comment:
Why are we not handling removal of resource from the state if Get returns a 404?

Reply:
How is this operation done in the new framework? We used to set the id to empty (d`.setID("")`) but I don't see something similar in the delete operation. https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1365/files#r1298282570

Reply:
Ah I found it https://developer.hashicorp.com/terraform/plugin/framework/migrating/resources/crud

> In SDKv2, calling d.SetId("") would signal resource removal. In the Framework, this is replaced with resp.State.RemoveResource(). Resource removal should only occur during the Read method to prevent Provider produced inconsistent result after apply errors from Terraform during other operations. The Delete method will automatically call resp.State.RemoveResource() if there are no errors.

Reply:
updated